icu4c/source/i18n/decimfmt.cpp

```diff
@@ -97,15 +101,17 @@ DecimalFormat::DecimalFormat(const UnicodeString& pattern, DecimalFormatSymbols*
 }
 
 DecimalFormat::DecimalFormat(const DecimalFormatSymbols* symbolsToAdopt, UErrorCode& status) {
+    // we must take ownership of symbolsToAdopt, even in a failure case.
     LocalPointer<const DecimalFormatSymbols> adoptedSymbols(symbolsToAdopt);
-    fields = new DecimalFormatFields();
     if (U_FAILURE(status)) {
         return;
     }
+    fields = new DecimalFormatFields();
     if (fields == nullptr) {
         status = U_MEMORY_ALLOCATION_ERROR;
         return;
     }
+    fields->formatter.adoptInsteadAndCheckErrorCode(new LocalizedNumberFormatter(), status);
```

Comment:
hmm.  It's not great how you have to check fFields (this->fields) and all of its children everywhere.  Besides being hard to read, it could be dangerous if a new child field is added to fFields and the checks get out of sync with the list of actual fields.

Three ideas of ways around this:

1. Add a method on the type DecimalFormatFields that returns a bool for whether the thing is valid or not, such that your check becomes `if (!fields || !fields->isValid()) { return; }`
2. Here and in other relevant constructors, if the allocation fails for any of the "children" of fFields, then call delete on fFields and set fFields to nullptr.  In other words, forbid any fFields unless it is completely valid.
3. Make the children of fFields be "stack"-allocated.  I made a follow-up ticket for this:

https://unicode-org.atlassian.net/browse/ICU-20394